File: python/src/azoth/eos/reference/phase_transport.py
"""``eos.phase_transport`` - a phase's transport properties, by NeqSim's phase-type dispatch.

Spec: ``specs/models/eos/phase_transport.toml``

The Python twin of ``crates/azoth-eos/src/phase_transport.rs``, written to mirror it line for
line: it composes the same reference kernels, in the same order, and resolves the same
component data.

# The dispatch is the answer

``PhysicalProperties`` is built by phase type and each subclass assigns its own models, so one
fluid's gas and liquid answer their properties from *different correlations*:

===========  ==========================  ==============================  ==================
phase        viscosity                   conductivity                    diffusivity
===========  ==========================  ==============================  ==================
gas          PFCT (heavy oil)            PFCT                            Chapman-Enskog
oil          PFCT                        PFCT                            Siddiqi-Lucas
aqueous      the polynom ``Viscosity``   the polynom ``Conductivity``     Siddiqi-Lucas
===========  ==========================  ==============================  ==================

# The two ladders, and the fallback that hides one of them

The liquid diffusivity divides by a **pure-component** viscosity, which comes from whichever
LIQVISC ladder the phase's viscosity class carries. An oil's PFCT class inherits the common-phase
one, whose model 2 branch is **empty** - and ``SiddiqiLucasMethod`` then falls back to the
*phase's own* viscosity, so the hole is the normal path rather than an error.
"""

from __future__ import annotations

from collections.abc import Sequence

from azoth.core.errors import InvalidInputError, OutOfRangeError
from azoth.core.range import apply_checks, checks_for
from azoth.core.result import PhaseTransportResult
from azoth.core.units import Q, from_si, input_to_si, ureg
from azoth.core.warnings import Warning
from azoth.eos.components import (
    entry,
    lennard_jones_pair,
    mixture_of,
    normal_boiling_molar_volume,
)
from azoth.eos.reference.aqueous_viscosity import aqueous_viscosity
from azoth.eos.reference.chapman_enskog_diffusivity import chapman_enskog_diffusivity
from azoth.eos.reference.effective_diffusion import effective_diffusion
from azoth.eos.reference.liquid_conductivity_polynom import liquid_conductivity_polynom
from azoth.eos.reference.liquid_viscosity_pure import liquid_viscosity_pure
from azoth.eos.reference.siddiqi_lucas_diffusivity import siddiqi_lucas_diffusivity
from azoth.eos.reference.thermal_conductivity import thermal_conductivity
from azoth.eos.reference.viscosity import viscosity

CALC_ID = "eos.phase_transport"

#: NeqSim's `PhaseType` values, as the spec spells them.
KINDS = ("gas", "oil", "aqueous")

#: `SiddiqiLucasMethod`'s own floor on the pure-component viscosity, in cP.
MIN_ETA_CP = 0.01

#: The ladder each liquid kind's viscosity class carries.
LADDERS = {"oil": "common_phase", "aqueous": "liquid"}


def phase_transport(
    components: Sequence[str],
    phase: str,
    T: Q,
    P: Q,
    z: Sequence[float],
) -> PhaseTransportResult:
    """A phase's viscosity, conductivity, binary diffusivity matrix and effective diffusivities.

    Args:
        components: the phase's substances, by name.
        phase: ``gas``, ``oil`` or ``aqueous`` - **which decides the correlations**.
        T: absolute temperature.
        P: absolute pressure.
        z: the phase's mole fractions.

    Returns:
        The four properties the segment model's transport snapshot reads.

    Raises:
        InvalidInputError: for a phase kind that is not one of the three, a composition of the
            wrong length, or a phase with fewer than two components - the effective diffusivity
            divides by an empty sum there and NeqSim answers a ``NaN``.
        OutOfRangeError: from whichever correlation refuses the state.

    Example:
        >>> import azoth
        >>> q = azoth.ureg.Quantity
        >>> r = phase_transport(
        ...     ["CO2", "water"], "aqueous", q(313.15, "K"), q(50.0e5, "Pa"),
        ...     [0.0006691762234084198, 0.9993308237765915])
        >>> round(r.k.magnitude, 13)
        0.6344057039895
    """
    from azoth._models_gen import model

    spec = model(CALC_ID)
    checks = checks_for(spec)
    warnings: list[Warning] = []

    if phase not in KINDS:
        raise InvalidInputError(
            "phase",
            f"`{phase}` is not one of NeqSim's three phase kinds: {', '.join(KINDS)}",
        )
    count = len(components)
    if len(z) != count:
        raise InvalidInputError(
            "z", f"a mixture of {count} component(s) needs {count} mole fractions, got {len(z)}"
        )
    if count < 2:
        raise InvalidInputError(
            "components",
            "one component: the effective diffusivity divides by the sum over the *other* "
            "components, so NeqSim answers a NaN here and this refuses instead of reporting one",
        )

    temperature = input_to_si(spec, "T", T)
    pressure = input_to_si(spec, "P", P)
    apply_checks(
        checks.on_input,
        {
            "T": temperature,
            "P": pressure,
            "z": min(float(value) for value in z),
        }.get,
        warnings,
    )

    mixture, ideal_gas = mixture_of(list(components))
    entries = [entry(name) for name in components]
    molar_mass: list[Q] = []
    for record in entries:
        if record.molar_mass is None:
            raise InvalidInputError(
                "components",
                "a component carries no molar mass, and every correlation here is a "
                "function of one",
            )
        molar_mass.append(record.molar_mass)

    # ---- The viscosity and the conductivity, by phase kind.
    if phase == "aqueous":
        mu = aqueous_viscosity(mixture, T, P, list(z)).viscosity
        k = liquid_conductivity_polynom(
            [record.liquid_conductivity for record in entries],
            molar_mass,
            list(z),
            T,
        ).k
    else:
        mu = viscosity(mixture, T, P, list(z)).mu
        k = thermal_conductivity(mixture, ideal_gas, T, P, list(z)).k

    # ---- The binary diffusivity matrix, by phase kind.
    matrix: list[list[Q]] = [[from_si(0.0, "m**2/s") for _ in range(count)] for _ in range(count)]
    for i in range(count):
        for j in range(count):
            if i == j:
                continue
            if phase == "gas":
                sigma, eps, _ = lennard_jones_pair(
                    entries[i].lennard_jones_diameter,
                    entries[i].lennard_jones_energy,
                    entries[j].lennard_jones_diameter,
                    entries[j].lennard_jones_energy,
                    molar_mass[i].to("kg/mol").magnitude,
                    molar_mass[j].to("kg/mol").magnitude,
                )
                matrix[i][j] = chapman_enskog_diffusivity(
                    molar_mass[i],
                    molar_mass[j],
                    ureg.Quantity(sigma * 1.0e-10, "m"),
                    ureg.Quantity(eps, "K"),
                    T,
                    P,
                ).d
            else:
                # **The aqueous form for every liquid**: the method's `autoSelectCorrelation` is
                # false, so an oil takes the aqueous correlation too.
                va = normal_boiling_molar_volume(
                    _liquid_density(entries[i]),
                    molar_mass[i].to("kg/mol").magnitude,
                    _critical_volume(entries[i]),
                )
                vb = normal_boiling_molar_volume(
                    _liquid_density(entries[j]),
                    molar_mass[j].to("kg/mol").magnitude,
                    _critical_volume(entries[j]),
                )
                # The pure-component viscosity the correlation divides by, from **this phase
                # kind's ladder** - and the class's own fallback where it is not positive, which
                # on an oil is the normal path: the common-phase ladder's model 2 branch is
                # empty, so the phase's own viscosity is what the correlation takes.
                pure = (
                    liquid_viscosity_pure(
                        LADDERS[phase],
                        entries[j].liqvisc_model,
                        *entries[j].liqvisc,
                        entries[j].Tc,
                        entries[j].Pc,
                        entries[j].omega,
                        T,
                        P,
                    )
                    .mu.to("Pa*s")
                    .magnitude
                    * 1000.0
                )
                eta_cp = pure if pure > 0.0 else mu.to("Pa*s").magnitude * 1000.0
                matrix[i][j] = siddiqi_lucas_diffusivity(
                    "aqueous",
                    ureg.Quantity(va * 1.0e-6, "m**3/mol"),
                    ureg.Quantity(vb * 1.0e-6, "m**3/mol"),
                    T,
                    ureg.Quantity(max(eta_cp, MIN_ETA_CP) * 1.0e-3, "Pa*s"),
                ).d

    # The assembly takes quantities, not magnitudes: it is the same kernel the id exposes.
    #
    # **A phase that is one substance has no effective diffusivity, and that is not this id's
    # refusal.** The assembly divides by the *other* components' fractions, so on a pure phase
    # there is nothing for the one component to diffuse into and ``eos.effective_diffusion``
    # refuses - rightly, for the question *it* asks. ``RateBasedPackedColumnTest``'s own lean
    # solvent is pure water, and NeqSim answers a **zero** vector there rather than refusing;
    # the Rust twin carries the same narrowing for the same measurement.
    try:
        assembled = effective_diffusion(matrix, list(z))
        d_effective = assembled.effective_diffusion
        warnings.extend(assembled.warnings)
    except OutOfRangeError as refusal:
        if refusal.input_field != "x":
            raise
        d_effective = tuple(from_si(0.0, "m**2/s") for _ in z)

    apply_checks(
        checks.derived,
        lambda name: (
            mu.to("Pa*s").magnitude
            if name == "mu"
            else (k.to("W/(m*K)").magnitude if name == "k" else None)
        ),
        warnings,
    )

    return PhaseTransportResult(
        mu=mu,
        k=k,
        d_binary=tuple(tuple(value for value in row) for row in matrix),
        d_effective=d_effective,
        warnings=tuple(warnings),
    )


def _liquid_density(record: object) -> float:
    """A component's normal liquid density in kg/m**3, zero where the row carries none.

    **Zero is the table's own spelling of absence** - `getNormalLiquidDensity` estimates from
    the critical volume there - so this passes it through rather than substituting one.
    """
    value = getattr(record, "liquid_density", None)
    return 0.0 if value is None else float(value.to("kg/m**3").magnitude)


def _critical_volume(record: object) -> float | None:
    """A component's critical volume in m**3/mol, `None` where the row carries none."""
    value = getattr(record, "critical_volume", None)
    return None if value is None else float(value.to("m**3/mol").magnitude)

File: crates/azoth-eos/src/phase_transport.rs
//! `eos.phase_transport` - a phase's transport properties, by the dispatch NeqSim's
//! `PhysicalProperties` makes.
//!
//! Spec: `specs/models/eos/phase_transport.toml`, which records the three phase kinds' model
//! selections and the two ladders the diffusivity chain reads.

use azoth_core::units::{
    Pressure, ThermodynamicTemperature, pascal_seconds, watts_per_meter_kelvin,
};
use azoth_core::{AzothError, Result, apply_checks};

use crate::IdealGasModel;
use crate::chapman_enskog_diffusivity::chapman_enskog_diffusivity;
use crate::databank::{lennard_jones_pair, normal_boiling_molar_volume};
use crate::effective_diffusion::effective_diffusion;
use crate::liquid_conductivity_polynom::liquid_conductivity_polynom;
use crate::liquid_viscosity_pure::{LiquidViscosityLadder, liquid_viscosity_pure};
use crate::mixture::Mixture;
use crate::model_gen;
use crate::results::PhaseTransportResult;
use crate::siddiqi_lucas_diffusivity::{SiddiqiLucasForm, siddiqi_lucas_diffusivity};
use crate::thermal_conductivity::thermal_conductivity;
use crate::viscosity::viscosity;

/// The three phase kinds `PhysicalProperties` dispatches on.
///
/// NeqSim's `PhaseType`, and the dispatch is by it: an aqueous phase takes `WaterPhysicalProperties`,
/// whose viscosity and conductivity are the *polynom* correlations, while a gas and a hydrocarbon
/// liquid take PFCT for both.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum PhaseKind {
    /// `PhaseType.GAS`: PFCT viscosity and conductivity, Chapman-Enskog diffusivity.
    Gas,
    /// `PhaseType.OIL`: PFCT for both, Siddiqi-Lucas diffusivity on the common-phase ladder.
    Oil,
    /// `PhaseType.AQUEOUS`: the polynom viscosity and conductivity, Siddiqi-Lucas diffusivity on
    /// the liquid ladder.
    Aqueous,
}

impl PhaseKind {
    /// The spec's spelling.
    #[must_use]
    pub const fn as_str(self) -> &'static str {
        match self {
            Self::Gas => "gas",
            Self::Oil => "oil",
            Self::Aqueous => "aqueous",
        }
    }
}

impl std::str::FromStr for PhaseKind {
    type Err = String;

    fn from_str(s: &str) -> std::result::Result<Self, Self::Err> {
        match s {
            "gas" => Ok(Self::Gas),
            "oil" => Ok(Self::Oil),
            "aqueous" => Ok(Self::Aqueous),
            other => Err(format!(
                "`{other}` is not one of NeqSim's three phase kinds: gas, oil or aqueous"
            )),
        }
    }
}

/// The lowest pure-component viscosity the liquid diffusivity correlations take, in cP.
///
/// `SiddiqiLucasMethod.calcBinaryDiffusionCoefficient` clamps `eta` to this, and the clamp is
/// reachable: the common-phase ladder answers **zero** for a component whose LIQVISC model is 2.
const MIN_ETA_CP: f64 = 0.01;

/// A phase's viscosity, thermal conductivity, binary diffusivity matrix and effective
/// diffusivities - the four the segment model's transport snapshot reads.
///
/// **The dispatch is the answer, not a detail.** NeqSim's `PhysicalProperties` is created by
/// phase type and each subclass assigns its own models, so the same fluid's gas and liquid answer
/// their properties from *different correlations*. The capture holds both the values and the model
/// classes per phase, which is what the cases are held to.
///
/// # Errors
/// * [`azoth_core::AzothError::InvalidInput`] if `z` is not one entry per component, or if the
///   phase has fewer than two components - the effective diffusivity divides by an empty sum
///   there and NeqSim answers a `NaN`.
/// * Whatever the composed correlations refuse.
pub fn phase_transport(
    mixture: &Mixture,
    ideal_gas: &IdealGasModel,
    phase: PhaseKind,
    t: ThermodynamicTemperature,
    p: Pressure,
    z: &[f64],
) -> Result<PhaseTransportResult> {
    let spec = &model_gen::PHASE_TRANSPORT_SPEC;
    let mut warnings = Vec::new();

    let components = mixture.components();
    let count = components.len();
    if z.len() != count {
        return Err(AzothError::invalid_input(
            "z",
            format!(
                "a mixture of {count} components needs {count} mole fractions, got {}",
                z.len()
            ),
        ));
    }
    if count < 2 {
        return Err(AzothError::invalid_input(
            "components",
            "one component: the effective diffusivity divides by the sum over the *other* \
             components, so NeqSim answers a NaN here and this refuses instead of reporting \
             one"
            .to_string(),
        ));
    }

    // Every correlation here is a function of the components' molar masses, and NeqSim
    // refuses a component that has none rather than inventing one.
    let molar_mass: Vec<f64> = components
        .iter()
        .map(|component| {
            component.molar_mass.ok_or_else(|| {
                AzothError::invalid_input(
                    "components",
                    "a component carries no molar mass, and every correlation here is a                      function of one",
                )
            })
        })
        .collect::<Result<Vec<f64>>>()?;

    apply_checks(
        spec.input_checks(),
        |quantity| match quantity {
            "T" => Some(t.value),
            "P" => Some(p.value),
            "z" => z.iter().copied().reduce(f64::min),
            _ => None,
        },
        &mut warnings,
    )?;

    // ---- The viscosity and the conductivity, by phase kind.
    let (mu, k) = match phase {
        PhaseKind::Aqueous => (
            crate::aqueous_viscosity::aqueous_viscosity(mixture, t, p, z)?
                .viscosity
                .value,
            liquid_conductivity_polynom(
                &components
                    .iter()
                    .map(|component| component.liquid_conductivity)
                    .collect::<Vec<[f64; 3]>>(),
                &molar_mass
                    .iter()
                    .map(|mass| azoth_core::units::kilograms_per_mole(*mass))
                    .collect::<Vec<_>>(),
                z,
                t,
            )?
            .k
            .value,
        ),
        PhaseKind::Gas | PhaseKind::Oil => (
            viscosity(mixture, t, p, z)?.mu.value,
            thermal_conductivity(mixture, ideal_gas, t, p, z)?.k.value,
        ),
    };

    // ---- The binary diffusivity matrix, by phase kind.
    let d_binary = match phase {
        PhaseKind::Gas => {
            let mut matrix = vec![vec![0.0; count]; count];
            for i in 0..count {
                for j in 0..count {
                    if i == j {
                        continue;
                    }
                    let (sigma, eps, _) = lennard_jones_pair(
                        components[i].lennard_jones_diameter,
                        components[i].lennard_jones_energy,
                        components[j].lennard_jones_diameter,
                        components[j].lennard_jones_energy,
                        molar_mass[i],
                        molar_mass[j],
                    );
                    matrix[i][j] = chapman_enskog_diffusivity(
                        azoth_core::units::kilograms_per_mole(molar_mass[i]),
                        azoth_core::units::kilograms_per_mole(molar_mass[j]),
                        sigma * 1.0e-10,
                        azoth_core::units::kelvins(eps),
                        t,
                        p,
                    )?
                    .d
                    .value;
                }
            }
            matrix
        }
        PhaseKind::Oil | PhaseKind::Aqueous => {
            // **The default is the aqueous form for every liquid.** `SiddiqiLucasMethod`'s
            // `autoSelectCorrelation` is false, and `setDiffusionCoefficientModel` is the only
            // thing that turns it on - so an oil takes the aqueous correlation here.
            //
            // **And the two phase kinds take different ladders** for the `eta` it divides by:
            // the oil phase's PFCT viscosity class inherits the common-phase one.
            let ladder = match phase {
                PhaseKind::Aqueous => LiquidViscosityLadder::Liquid,
                _ => LiquidViscosityLadder::CommonPhase,
            };
            let mut matrix = vec![vec![0.0; count]; count];
            for i in 0..count {
                for j in 0..count {
                    if i == j {
                        continue;
                    }
                    let va = normal_boiling_molar_volume(
                        components[i].normal_liquid_density,
                        molar_mass[i],
                        components[i].critical_volume,
                    );
                    let vb = normal_boiling_molar_volume(
                        components[j].normal_liquid_density,
                        molar_mass[j],
                        components[j].critical_volume,
                    );
                    let pure_cp = liquid_viscosity_pure(
                        ladder,
                        components[j].liqvisc_model,
                        components[j].liqvisc[0],
                        components[j].liqvisc[1],
                        components[j].liqvisc[2],
                        components[j].liqvisc[3],
                        components[j].tc,
                        components[j].pc,
                        components[j].omega,
                        t,
                        p,
                    )?
                    .mu
                    .value
                        * 1000.0;
                    // **The class's own fallback, and on this path it is the normal one**: a
                    // pure-component viscosity that is not positive - exactly what the
                    // common-phase ladder's empty model 2 branch answers - takes the *phase's*
                    // viscosity instead. Measured on the captured oil phase, where the hole
                    // fires and `eta` is the phase's own `0.1347` cP rather than the `0.01`
                    // the clamp would otherwise floor it to.
                    let eta_cp = if pure_cp.is_finite() && pure_cp > 0.0 {
                        pure_cp
                    } else {
                        mu * 1000.0
                    };
                    matrix[i][j] = siddiqi_lucas_diffusivity(
                        SiddiqiLucasForm::Aqueous,
                        azoth_core::units::cubic_meters_per_mole(va * 1.0e-6),
                        azoth_core::units::cubic_meters_per_mole(vb * 1.0e-6),
                        t,
                        pascal_seconds(eta_cp.max(MIN_ETA_CP) * 1.0e-3),
                    )?
                    .d
                    .value;
                }
            }
            matrix
        }
    };

    // **A phase that is one substance has no effective diffusivity, and that is not this
    // id's refusal.** The assembly divides by the *other* components' fractions, so on a
    // pure phase there is nothing for the one component to diffuse into and
    // `eos.effective_diffusion` refuses - rightly, for the question *it* asks.
    // `RateBasedPackedColumnTest`'s own lean solvent is pure water, and NeqSim answers a
    // **zero** vector there rather than refusing: measured, the capture's absorber row prints
    // `0.0` for every entry of both phases. So the vector is zeros, and the three properties
    // that *are* defined - the viscosity, the conductivity and the pair matrix - are reported.
    let d_effective = match effective_diffusion(&d_binary, z) {
        Ok(assembled) => {
            warnings.extend(assembled.warnings.iter().cloned());
            assembled.effective_diffusion
        }
        Err(AzothError::OutOfRange { ref field, .. }) if field == "x" => vec![0.0; count],
        Err(other) => return Err(other),
    };

    apply_checks(
        spec.derived_checks(),
        |name| match name {
            "mu" => Some(mu),
            "k" => Some(k),
            _ => None,
        },
        &mut warnings,
    )?;

    Ok(PhaseTransportResult {
        mu: pascal_seconds(mu),
        k: watts_per_meter_kelvin(k),
        d_binary,
        d_effective,
        warnings,
    })
}
